include/sys/onoff.h

```diff
@@ -0,0 +1,611 @@
+ *
+ * Objects of this type are allocated by a client, which must use an
+ * initialization function (e.g. onoff_client_init_signal()) to
+ * configure them.
+ *
+ * Control of the object content transfers to the service provider
+ * when a pointer to the object is passed to any on-off service
+ * function.  While the service provider controls the object the
+ * client must not change any object fields.  Control reverts to the
+ * client:
+ * * if the call to the service API returns an error;
+ * * if the call to the service API succeeds for a no-wait operation;
+ * * when operation completion is posted (signalled or callback
+ *   invoked).
+ *
+ * Only the result field is intended for direct use by clients, and it
+ * is available for inspection only after control reverts to the
+ * client.
+ */
+struct onoff_client {
+	/* Links the client into the set of waiting service users. */
+	sys_snode_t node;
+
+	union async {
+		/* Pointer to signal used to notify client.
+		 *
+		 * The signal value corresponds to the res parameter
+		 * of onoff_client_callback.
+		 */
+#ifdef CONFIG_POLL
+		struct k_poll_signal *signal;
+#endif /* CONFIG_POLL */
+
+		/* Handler and argument for callback notification. */
+		struct callback {
+			onoff_client_callback handler;
+			void *user_data;
+		} callback;
+	} async;
+
+	/*
+	 * The result of the operation.
+	 *
+	 * This is the value that was (or would be) passed to the
+	 * async infrastructure.  This field is the sole record of
+	 * success or failure for no-wait synchronous operations.
+	 */
+	int volatile result;
+
+	/* Flags recording client state. */
+	unsigned int volatile flags;
+};
+
+/**
+ * @brief Check for and read the result of an asynchronous operation.
+ *
+ * @param op pointer to the object used to specify asynchronous
+ * function behavior and store completion information.
+ *
+ * @param result pointer to storage for the result of the operation.
+ * The result is stored only if the operation has completed.
+ *
+ * @retval 0 if the operation has completed.
+ * @retval -EAGAIN if the operation has not completed.
+ */
+static inline int onoff_client_fetch_result(const struct onoff_client *op,
+					    int *result)
+{
+	__ASSERT_NO_MSG(op != NULL);
+	__ASSERT_NO_MSG(result != NULL);
+
+	int rv = -EAGAIN;
+
+	if (op->flags == 0U) {
+		rv = 0;
+		*result = op->result;
+	}
+	return rv;
+}
+
+/**
+ * @brief Initialize an on-off client to be used for a spin-wait
+ * operation notification.
+ *
+ * Clients that use this initialization receive no asynchronous
+ * notification, and instead must periodically check for completion
+ * using onoff_client_fetch_result().
+ *
+ * @param cli pointer to the client state object.
+ */
+static inline void onoff_client_init_spinwait(struct onoff_client *cli)
+{
+	__ASSERT_NO_MSG(cli != NULL);
+
+	*cli = (struct onoff_client){
+		.flags = ONOFF_CLIENT_NOTIFY_SPINWAIT,
+	};
+}
+
+#if defined(CONFIG_POLL) || defined(DOXYGEN)
```

Comment:
the better solution is not to enclose APIs in #ifdefs and instead document that a certain API requires a Kconfig to be enabled, something like that...

```
 * @note
 *    @rst
 *    You should enable :option:`CONFIG_POLL` in your project
 *    configuration.
 *    @endrst
```

I am not a fan of adding this check for DOXYGEN. We are moving the logic from a simple list in doxyfile into the code which is a very gross workaround and might cause confusion and in some cases will not be as simple as here, depending on the logic.

Reply:
> We are moving the logic from a simple list in doxyfile into the code which is a very gross workaround and might cause confusion and in some cases will not be as simple as here, depending on the logic.

I don't agree that the doxyfile list is "simple". Adding CONFIG_FOO=y to it when we want to document some new API could result in unexpected changes elsewhere if the author expected some CONFIG_  defines to be unset when they placed their docstrings.